Plugins/GitUIPluginInterfaces/ManagedExtensibility.cs

```diff
@@ -1,97 +1,91 @@
 ﻿using System;
 using System.Collections.Generic;
-using System.ComponentModel.Composition.Hosting;
 using System.Diagnostics;
 using System.IO;
 using System.Linq;
 using System.Reflection;
+using System.Threading;
 using System.Windows.Forms;
+using GitUI;
+using Microsoft.VisualStudio.Composition;
 
 namespace GitUIPluginInterfaces
 {
     public static class ManagedExtensibility
     {
-        private static List<CompositionContainer> _compositionContainers;
+        private static readonly Lazy<ExportProvider> _exportProvider = new Lazy<ExportProvider>(CreateExportProvider, LazyThreadSafetyMode.ExecutionAndPublication);
 
-        private static readonly object compositionContainerSyncObj = new object();
+        private static ExportProvider CreateExportProvider()
+        {
+            var stopwatch = Stopwatch.StartNew();
 
-        private static readonly HashSet<string> _loggedExceptionMessages = new HashSet<string>();
+            var file = new FileInfo(Application.ExecutablePath);
 
-        /// <summary>
-        /// The MEF container.
-        /// </summary>
-        private static List<CompositionContainer> GetCompositionContainers()
-        {
-            lock (compositionContainerSyncObj)
-            {
-                if (_compositionContainers == null)
-                {
-                    _compositionContainers = new List<CompositionContainer>();
-                    var pluginsDir = new DirectoryInfo(Directory.GetParent(Application.ExecutablePath).FullName + Path.DirectorySeparatorChar + "Plugins");
-                    if (pluginsDir.Exists)
-                    {
-                        foreach (var dll in pluginsDir.EnumerateFiles("*.dll"))
-                        {
-                            _compositionContainers.Add(new CompositionContainer(new DirectoryCatalog(dll.DirectoryName, dll.Name)));
-                        }
-                    }
-                }
+            FileInfo[] plugins =
+                Directory.Exists(Path.Combine(file.Directory.FullName, "Plugins"))
+                ? new DirectoryInfo(Path.Combine(file.Directory.FullName, "Plugins")).GetFiles("*.dll")
+                : new FileInfo[] { };
 
-                return _compositionContainers;
-            }
-        }
+            var pluginFiles = plugins;
 
-        public static IEnumerable<Lazy<T>> GetExports<T>()
-        {
-            var ret = new List<Lazy<T>>();
-            foreach (var container in GetCompositionContainers())
+            var cacheFile = @"Plugins\GitUIPluginInterfaces\composition.cache";
```

Comment:
❓ This should go somewhere under `%LOCALAPPDATA%`. Is there a clean was to determine the best path for the cache?

Reply:
I think it should go somewhere under AppSettings.ApplicationDataPath. Perhaps ManagedExtensibility should have it injected.

Reply:
This works for me. As for clearing the cache, I was planning to use the cache if it's available and mark it as invalid asynchronously via a background operation that runs after it loads. A second alternative is to only clear the cache via manual action, which the plugin manager could do whenever it modified the set of installed plugins.